File: envelope.go
// Package pubsub es el cliente compartido de Pub/Sub de Discordia sobre
// RabbitMQ (INF-06). Expone una API única de publish/subscribe para que
// ningún servicio tenga que resolver por su cuenta el envelope de eventos,
// la topología de colas, el ack manual, los reintentos ni la reconexión.
package pubsub

import (
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"time"

	"github.com/google/uuid"
)

// DefaultEventVersion es la versión con la que nace todo evento nuevo. Un
// evento existente nunca cambia de forma: si el payload tiene que cambiar de
// manera incompatible se publica la versión siguiente en paralelo y se
// deprecia la anterior.
const DefaultEventVersion = 1

// eventTypePattern implementa el naming `<servicio>.<agregado>.<evento-en-pasado>`
// de las convenciones del equipo (ej. `chat.message.sent`). Son exactamente
// tres segmentos en minúscula: los bindings de los consumers dependen de esa
// forma (`community.#`, `*.member.banned`), así que un evento con otra forma
// no le llega a quien debería y conviene rechazarlo al construirlo.
var eventTypePattern = regexp.MustCompile(`^[a-z][a-z0-9]*(?:-[a-z0-9]+)*(?:\.[a-z][a-z0-9]*(?:-[a-z0-9]+)*){2}$`)

// Envelope es el sobre común obligatorio de todo evento del bus, definido en
// INF-02 (`discordia-docs/arquitectura/eventos.md`). Los campos van en
// camelCase porque el mismo formato viaja hasta el front.
type Envelope struct {
	// EventID identifica unívocamente esta publicación. Es la clave de
	// idempotencia: un consumer que ya procesó este EventID tiene que poder
	// descartarlo sin efectos.
	EventID string `json:"eventId"`

	// EventType es el nombre del evento y, a la vez, la routing key con la
	// que se publica en el exchange.
	EventType string `json:"eventType"`

	// EventVersion permite convivir dos formas del mismo evento durante una
	// migración.
	EventVersion int `json:"eventVersion"`

	// OccurredAt es cuándo ocurrió el hecho, en UTC. Es la única fuente de
	// verdad sobre qué es más nuevo: el bus no garantiza orden de llegada.
	OccurredAt time.Time `json:"occurredAt"`

	// CorrelationID es el hilo que une todos los eventos de una misma acción
	// del usuario, de punta a punta entre servicios.
	CorrelationID string `json:"correlationId"`

	// CausationID es el EventID del evento que provocó este. Vacío si el
	// evento nace de una acción directa y no de otro evento.
	CausationID string `json:"causationId,omitempty"`

	// Producer es el servicio que publicó el evento (`chat`, `identity`, …).
	Producer string `json:"producer"`

	// Data es el payload propio del evento. Se deja crudo para que cada
	// consumer lo deserialice en su propio tipo con UnmarshalData.
	Data json.RawMessage `json:"data"`
}

// EnvelopeOption ajusta un Envelope en construcción.
type EnvelopeOption func(*Envelope)

// WithCorrelationID fija el correlation ID en vez de generar uno nuevo. Es lo
// que hay que usar cuando la acción ya viene con uno desde el api-gateway.
func WithCorrelationID(id string) EnvelopeOption {
	return func(e *Envelope) { e.CorrelationID = id }
}

// WithCause marca este evento como consecuencia de otro: hereda su
// correlation ID y guarda su EventID como causation ID. Es la forma de que
// una cadena de reacciones asíncronas siga siendo trazable.
func WithCause(parent Envelope) EnvelopeOption {
	return func(e *Envelope) {
		e.CorrelationID = parent.CorrelationID
		e.CausationID = parent.EventID
	}
}

// WithEventVersion publica una versión distinta de la inicial del evento.
func WithEventVersion(version int) EnvelopeOption {
	return func(e *Envelope) { e.EventVersion = version }
}

// WithOccurredAt fija el momento del hecho cuando no es "ahora" — por
// ejemplo al republicar algo ya ocurrido, o en tests deterministas.
func WithOccurredAt(t time.Time) EnvelopeOption {
	return func(e *Envelope) { e.OccurredAt = t.UTC() }
}

// NewEnvelope arma un evento listo para publicar: genera EventID y
// correlation ID, sella OccurredAt en UTC, serializa data y valida el
// resultado. Devuelve error antes de tocar el bus si algo no cumple el
// contrato.
func NewEnvelope(eventType, producer string, data any, opts ...EnvelopeOption) (Envelope, error) {
	payload, err := json.Marshal(data)
	if err != nil {
		return Envelope{}, fmt.Errorf("marshal event data: %w", err)
	}

	envelope := Envelope{
		EventID:       uuid.NewString(),
		EventType:     eventType,
		EventVersion:  DefaultEventVersion,
		OccurredAt:    time.Now().UTC(),
		CorrelationID: uuid.NewString(),
		Producer:      producer,
		Data:          payload,
	}

	for _, opt := range opts {
		opt(&envelope)
	}

	if err := envelope.Validate(); err != nil {
		return Envelope{}, err
	}

	return envelope, nil
}

// ErrInvalidEnvelope encabeza todo error de validación del sobre, para que un
// consumer pueda distinguir "el mensaje está mal formado" de "falló el
// handler".
var ErrInvalidEnvelope = errors.New("invalid envelope")

// Validate verifica que el sobre cumpla el contrato de INF-02.
func (e Envelope) Validate() error {
	if _, err := uuid.Parse(e.EventID); err != nil {
		return fmt.Errorf("%w: eventId must be a UUID, got %q", ErrInvalidEnvelope, e.EventID)
	}
	if !eventTypePattern.MatchString(e.EventType) {
		return fmt.Errorf("%w: eventType %q must be <service>.<aggregate>.<past-tense-event>", ErrInvalidEnvelope, e.EventType)
	}
	if e.EventVersion < 1 {
		return fmt.Errorf("%w: eventVersion must be >= 1, got %d", ErrInvalidEnvelope, e.EventVersion)
	}
	if e.OccurredAt.IsZero() {
		return fmt.Errorf("%w: occurredAt is required", ErrInvalidEnvelope)
	}
	if _, err := uuid.Parse(e.CorrelationID); err != nil {
		return fmt.Errorf("%w: correlationId must be a UUID, got %q", ErrInvalidEnvelope, e.CorrelationID)
	}
	if e.CausationID != "" {
		if _, err := uuid.Parse(e.CausationID); err != nil {
			return fmt.Errorf("%w: causationId must be a UUID when present, got %q", ErrInvalidEnvelope, e.CausationID)
		}
	}
	if e.Producer == "" {
		return fmt.Errorf("%w: producer is required", ErrInvalidEnvelope)
	}
	if !isJSONObject(e.Data) {
		return fmt.Errorf("%w: data must be a JSON object", ErrInvalidEnvelope)
	}
	return nil
}

// RoutingKey es la clave con la que el evento se publica en el topic
// exchange. Coincide con el EventType por decisión de ADR-0003: así el
// binding de cada consumer se lee igual que el nombre del evento.
func (e Envelope) RoutingKey() string { return e.EventType }

// UnmarshalData deserializa el payload del evento en v.
func (e Envelope) UnmarshalData(v any) error {
	if err := json.Unmarshal(e.Data, v); err != nil {
		return fmt.Errorf("unmarshal data of event %s: %w", e.EventType, err)
	}
	return nil
}

// isJSONObject exige que el payload sea un objeto y no un escalar ni null:
// un evento cuyo data es `5` o `null` no se puede extender después sin
// romper a sus consumers.
func isJSONObject(raw json.RawMessage) bool {
	var probe map[string]any
	return len(raw) > 0 && json.Unmarshal(raw, &probe) == nil && probe != nil
}

File: python/discordia_pubsub/envelope.py
"""Sobre común de todo evento del bus, definido en INF-02.

Es el mismo contrato que implementa el cliente Go: los nombres de los campos
en el JSON son idénticos, así que un evento publicado por `identity` (Python)
lo consume `chat` (Go) sin traducción de por medio.
"""

from __future__ import annotations

import json
import re
import uuid
from dataclasses import dataclass, replace
from datetime import datetime, timezone
from typing import Any

# DEFAULT_EVENT_VERSION es la versión con la que nace todo evento nuevo. Un
# evento existente nunca cambia de forma: si el payload tiene que cambiar de
# manera incompatible se publica la versión siguiente en paralelo y se
# deprecia la anterior.
DEFAULT_EVENT_VERSION = 1

# El naming es `<servicio>.<agregado>.<evento-en-pasado>` (ej.
# `identity.user.registered`). Son exactamente tres segmentos en minúscula:
# los bindings de los consumers dependen de esa forma, así que un evento con
# otra forma no le llega a quien debería y conviene rechazarlo al construirlo.
_EVENT_TYPE_PATTERN = re.compile(r"^[a-z][a-z0-9]*(?:-[a-z0-9]+)*(?:\.[a-z][a-z0-9]*(?:-[a-z0-9]+)*){2}$")

# Go serializa los timestamps con precisión de nanosegundos y
# datetime.fromisoformat solo acepta hasta microsegundos. Recortar los
# dígitos de más es lo que hace que un evento publicado desde Go se pueda
# leer desde Python.
_TIMESTAMP_PATTERN = re.compile(r"^(.*?)(\.\d+)?(Z|[+-]\d{2}:?\d{2})$")


class InvalidEnvelopeError(ValueError):
    """El sobre no cumple el contrato de eventos."""


@dataclass(frozen=True, slots=True)
class Envelope:
    """Un evento del bus.

    Es inmutable a propósito: un evento describe algo que ya ocurrió, y un
    handler que lo modificara estaría cambiando el pasado para los demás
    handlers de su mismo proceso.
    """

    event_id: str
    event_type: str
    event_version: int
    occurred_at: datetime
    correlation_id: str
    producer: str
    data: dict[str, Any]
    causation_id: str | None = None

    @property
    def routing_key(self) -> str:
        """Clave con la que el evento se publica en el topic exchange.

        Coincide con el event_type por decisión de ADR-0003: así el binding
        de cada consumer se lee igual que el nombre del evento.
        """
        return self.event_type

    def validate(self) -> None:
        """Verifica que el sobre cumpla el contrato de INF-02."""
        _require_uuid(self.event_id, "eventId")
        if not _EVENT_TYPE_PATTERN.match(self.event_type or ""):
            raise InvalidEnvelopeError(
                f"eventType {self.event_type!r} must be <service>.<aggregate>.<past-tense-event>"
            )
        if not isinstance(self.event_version, int) or isinstance(self.event_version, bool) or self.event_version < 1:
            raise InvalidEnvelopeError(f"eventVersion must be an integer >= 1, got {self.event_version!r}")
        if not isinstance(self.occurred_at, datetime) or self.occurred_at.tzinfo is None:
            raise InvalidEnvelopeError("occurredAt must be a timezone-aware datetime")
        _require_uuid(self.correlation_id, "correlationId")
        if self.causation_id is not None:
            _require_uuid(self.causation_id, "causationId")
        if not self.producer:
            raise InvalidEnvelopeError("producer is required")
        if not isinstance(self.data, dict):
            raise InvalidEnvelopeError("data must be a JSON object")

    def to_dict(self) -> dict[str, Any]:
        """Serializa el sobre con las claves en camelCase del contrato."""
        payload: dict[str, Any] = {
            "eventId": self.event_id,
            "eventType": self.event_type,
            "eventVersion": self.event_version,
            "occurredAt": _format_timestamp(self.occurred_at),
            "correlationId": self.correlation_id,
        }
        # Mismo orden de claves que el cliente Go: los dos lados emiten un
        # JSON que se lee igual, y los fixtures de interoperabilidad se
        # comparan de un vistazo.
        if self.causation_id:
            payload["causationId"] = self.causation_id
        payload["producer"] = self.producer
        payload["data"] = self.data
        return payload

    def to_json(self) -> bytes:
        return json.dumps(self.to_dict()).encode()

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> Envelope:
        if not isinstance(payload, dict):
            raise InvalidEnvelopeError("an event must be a JSON object")
        try:
            return cls(
                event_id=payload["eventId"],
                event_type=payload["eventType"],
                event_version=payload["eventVersion"],
                occurred_at=_parse_timestamp(payload["occurredAt"]),
                correlation_id=payload["correlationId"],
                producer=payload["producer"],
                data=payload["data"],
                causation_id=payload.get("causationId"),
            )
        except KeyError as exc:
            raise InvalidEnvelopeError(f"missing field {exc.args[0]!r}") from exc

    @classmethod
    def from_json(cls, raw: bytes | str) -> Envelope:
        try:
            payload = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise InvalidEnvelopeError(f"event is not valid JSON: {exc}") from exc
        return cls.from_dict(payload)


def new_envelope(
    event_type: str,
    producer: str,
    data: dict[str, Any],
    *,
    correlation_id: str | None = None,
    cause: Envelope | None = None,
    event_version: int = DEFAULT_EVENT_VERSION,
    occurred_at: datetime | None = None,
) -> Envelope:
    """Arma un evento listo para publicar y lo valida.

    `cause` marca este evento como consecuencia de otro: hereda su
    correlation ID y guarda su event ID como causation ID. Es la forma de que
    una cadena de reacciones asíncronas siga siendo trazable.
    """
    if cause is not None:
        correlation_id = correlation_id or cause.correlation_id

    envelope = Envelope(
        event_id=str(uuid.uuid4()),
        event_type=event_type,
        event_version=event_version,
        occurred_at=(occurred_at or datetime.now(timezone.utc)).astimezone(timezone.utc),
        correlation_id=correlation_id or str(uuid.uuid4()),
        producer=producer,
        data=data,
        causation_id=cause.event_id if cause is not None else None,
    )
    envelope.validate()
    return envelope


def with_data(envelope: Envelope, data: dict[str, Any]) -> Envelope:
    """Devuelve una copia del sobre con otro payload, sin tocar el original."""
    return replace(envelope, data=data)


def _require_uuid(value: Any, field: str) -> None:
    try:
        uuid.UUID(str(value))
    except (ValueError, AttributeError, TypeError) as exc:
        raise InvalidEnvelopeError(f"{field} must be a UUID, got {value!r}") from exc


def _format_timestamp(value: datetime) -> str:
    """Fechas en UTC ISO-8601 con Z, como las emite el cliente Go."""
    return value.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


def _parse_timestamp(raw: Any) -> datetime:
    if isinstance(raw, datetime):
        return raw
    if not isinstance(raw, str):
        raise InvalidEnvelopeError(f"occurredAt must be an ISO-8601 string, got {raw!r}")

    match = _TIMESTAMP_PATTERN.match(raw)
    if match is None:
        raise InvalidEnvelopeError(f"occurredAt {raw!r} is not an ISO-8601 timestamp")

    head, fraction, offset = match.groups()
    # Go emite hasta nanosegundos; datetime solo entiende microsegundos.
    fraction = (fraction or "")[:7]
    normalized = f"{head}{fraction}{'+00:00' if offset == 'Z' else offset}"

    try:
        return datetime.fromisoformat(normalized).astimezone(timezone.utc)
    except ValueError as exc:
        raise InvalidEnvelopeError(f"occurredAt {raw!r} is not an ISO-8601 timestamp") from exc
